// @utcp/dotenv-loader/dist/index.js
// src/index.ts
import { VariableLoaderSerializer, ensureCorePluginsInitialized } from "@utcp/sdk";

// src/dotenv_variable_loader.ts
import { z } from "zod";
import { Serializer } from "@utcp/sdk";
import { promises as fs } from "fs";
import { parse as parseDotEnv } from "dotenv";
import * as path from "path";
var DotEnvVariableLoaderConfigSchema = z.object({
  variable_loader_type: z.literal("dotenv"),
  env_file_path: z.string().describe("Path to the .env file to load variables from.")
}).passthrough();
var DotEnvVariableLoader = class {
  variable_loader_type = "dotenv";
  env_file_path;
  constructor(env_file_path) {
    this.env_file_path = env_file_path;
  }
  /**
   * Retrieves a variable value from the .env file.
   * The file is read on every call to support runtime changes.
   * @param key Variable name to retrieve.
   * @returns Variable value if found, null otherwise.
   */
  async get(key) {
    try {
      const envFilePath = path.resolve(process.cwd(), this.env_file_path);
      const envContent = await fs.readFile(envFilePath, "utf-8");
      const envVars = parseDotEnv(envContent);
      return envVars[key] ?? null;
    } catch (e) {
      return null;
    }
  }
};
var DotEnvVariableLoaderSerializer = class extends Serializer {
  toDict(obj) {
    return {
      variable_loader_type: obj.variable_loader_type,
      env_file_path: obj.env_file_path
    };
  }
  validateDict(obj) {
    try {
      const validated = new DotEnvVariableLoader(DotEnvVariableLoaderConfigSchema.parse(obj).env_file_path);
      return validated;
    } catch (e) {
      if (e instanceof z.ZodError) {
        throw new Error(`Invalid DotEnvVariableLoader configuration: ${e.message}`);
      }
      throw new Error("Unexpected error during validation");
    }
  }
};

// src/index.ts
function register(override = false) {
  ensureCorePluginsInitialized();
  VariableLoaderSerializer.registerVariableLoader("dotenv", new DotEnvVariableLoaderSerializer(), override);
}
register();
export {
  DotEnvVariableLoader,
  DotEnvVariableLoaderSerializer,
  register
};
