// @utcp/dotenv-loader/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  DotEnvVariableLoader: () => DotEnvVariableLoader,
  DotEnvVariableLoaderSerializer: () => DotEnvVariableLoaderSerializer,
  register: () => register
});
module.exports = __toCommonJS(index_exports);
var import_sdk2 = require("@utcp/sdk");

// src/dotenv_variable_loader.ts
var import_zod = require("zod");
var import_sdk = require("@utcp/sdk");
var import_fs = require("fs");
var import_dotenv = require("dotenv");
var path = __toESM(require("path"), 1);
var DotEnvVariableLoaderConfigSchema = import_zod.z.object({
  variable_loader_type: import_zod.z.literal("dotenv"),
  env_file_path: import_zod.z.string().describe("Path to the .env file to load variables from.")
}).passthrough();
var DotEnvVariableLoader = class {
  variable_loader_type = "dotenv";
  env_file_path;
  constructor(env_file_path) {
    this.env_file_path = env_file_path;
  }
  /**
   * Retrieves a variable value from the .env file.
   * The file is read on every call to support runtime changes.
   * @param key Variable name to retrieve.
   * @returns Variable value if found, null otherwise.
   */
  async get(key) {
    try {
      const envFilePath = path.resolve(process.cwd(), this.env_file_path);
      const envContent = await import_fs.promises.readFile(envFilePath, "utf-8");
      const envVars = (0, import_dotenv.parse)(envContent);
      return envVars[key] ?? null;
    } catch (e) {
      return null;
    }
  }
};
var DotEnvVariableLoaderSerializer = class extends import_sdk.Serializer {
  toDict(obj) {
    return {
      variable_loader_type: obj.variable_loader_type,
      env_file_path: obj.env_file_path
    };
  }
  validateDict(obj) {
    try {
      const validated = new DotEnvVariableLoader(DotEnvVariableLoaderConfigSchema.parse(obj).env_file_path);
      return validated;
    } catch (e) {
      if (e instanceof import_zod.z.ZodError) {
        throw new Error(`Invalid DotEnvVariableLoader configuration: ${e.message}`);
      }
      throw new Error("Unexpected error during validation");
    }
  }
};

// src/index.ts
function register(override = false) {
  (0, import_sdk2.ensureCorePluginsInitialized)();
  import_sdk2.VariableLoaderSerializer.registerVariableLoader("dotenv", new DotEnvVariableLoaderSerializer(), override);
}
register();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DotEnvVariableLoader,
  DotEnvVariableLoaderSerializer,
  register
});
