// @hediet/node-reload/dist/FileWatcher.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileWatcher = void 0;
const fs_1 = require("fs");
const promises_1 = require("fs/promises");
const utils_1 = require("./utils");
class FileWatcher {
    constructor(_handleChanges) {
        this._handleChanges = _handleChanges;
        this._watchers = new Map();
        this._pendingFileWatchers = new Set();
        this._changedWatchers = new Set();
    }
    addFile(filename) {
        if (this._watchers.get(filename)) {
            throw new Error(`Tracker for ${filename} already set!`);
        }
        const q = new utils_1.AsyncQueue();
        const w = new SingleFileWatcher(filename, async () => {
            this._pendingFileWatchers.add(w);
            q.clear();
            q.schedule(() => wait(100));
            q.schedule(async () => {
                try {
                    const { didChange } = await w.update();
                    if (didChange) {
                        this._changedWatchers.add(w);
                    }
                }
                catch (e) {
                    console.error('unhandled error during update check', e);
                }
            });
            q.schedule(async () => {
                this._pendingFileWatchers.delete(w);
                if (this._pendingFileWatchers.size === 0) {
                    const filenames = Array.from(this._changedWatchers).map(w => w.filename);
                    this._changedWatchers.clear();
                    if (filenames.length > 0) {
                        this._handleChanges(filenames);
                    }
                }
            });
        });
        this._watchers.set(filename, w);
        return {
            dispose: () => {
                this._watchers.delete(filename);
                w.dispose();
            }
        };
    }
}
exports.FileWatcher = FileWatcher;
function wait(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}
class SingleFileWatcher {
    constructor(filename, _handleChange) {
        this.filename = filename;
        this._handleChange = _handleChange;
        this._fileContent = undefined;
        this._handler = () => {
            this._handleChange();
        };
        this._init = (async () => {
            const content = await (0, promises_1.readFile)(filename, 'utf-8');
            this._fileContent = content;
        })();
        this._watcher = (0, fs_1.watch)(filename, { persistent: false });
        this._watcher.on('change', this._handler);
    }
    dispose() {
        this._watcher.close();
    }
    async update() {
        await this._init;
        const content = await (0, promises_1.readFile)(this.filename, 'utf-8');
        const didChange = content !== this._fileContent;
        this._fileContent = content;
        return { didChange };
    }
}
