// @hediet/node-reload/dist/nodeApi.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getLoadedModule = getLoadedModule;
exports.deleteModule = deleteModule;
exports.resolveFileName = resolveFileName;
exports.moduleFromNodeModule = moduleFromNodeModule;
exports.registerModuleInterceptors = registerModuleInterceptors;
const Module = require("module");
const nodeJsModuleBrand = Symbol("NodeJsModule");
function getLoadedModule(filename) {
    return require.cache[filename];
}
function deleteModule(filename) {
    if (!require.cache[filename]) {
        return false;
    }
    delete require.cache[filename];
    return true;
}
function resolveFileName(request, caller) {
    return Module._resolveFilename(request, caller);
}
function moduleFromNodeModule(nodeModule) {
    return nodeModule;
}
function registerModuleInterceptors(options) {
    const originalModule = {
        load: Module.prototype.load,
        require: Module.prototype.require,
    };
    let disposed = false;
    const newRequire = Module.prototype.require = function (request) {
        if (disposed) {
            return originalModule.require.call(this, request);
        }
        return options.interceptRequire(this, request);
    };
    const newLoad = Module.prototype.load = function (filename) {
        if (disposed) {
            return originalModule.load.call(this, filename);
        }
        return options.interceptLoad(this, filename);
    };
    return {
        originalRequire: (module, request) => originalModule.require.call(module, request),
        originalLoad: (module, filename) => originalModule.load.call(module, filename),
        dispose: () => {
            disposed = true;
            if (Module.prototype.require === newRequire) {
                Module.prototype.require = originalModule.require;
            }
            if (Module.prototype.load === newLoad) {
                Module.prototype.load = originalModule.load;
            }
        }
    };
}
